### ISSUE TITLE
Setting IPv4 flags affect fragment offset on OSX

### ISSUE THREAD

**Issue by faern** [libpnet/libpnet#164]
I have noticed that calling `pnet::packet::ipv4::MutableIpv4Packet::set_flags` does not affect the flag in the header, but rather the fragment offset on OSX 10.11. These headers are adjacent and I suspect that some bit manipulation is slightly off here.
I have not inspected this in more detail. Any more info I can provide?

**Comment by mrmonday** [comment]
Could you tell me a little more about how you're experiencing this? Do you have a test case I could try? I can't reproduce this locally - everything seems to work as expected for me.

**Comment by faern** [comment]
Sure, of course. I was quite tired yesterday, so I didn't have the energy to explain better, sorry for that.

Now I have a small program that will send two packets on layer 3 and I inspect their content on the wire with wireshark. On the first packet I never call `set_flags` and when I observe the packet in wireshark I see that the three flag bits are zero and fragment offset is also zero. On the second packet I call `set_flags(2)` but in wireshark the three flag bits are still 0 and fragment offset is now 512

```rust
extern crate pnet;

use std::net::Ipv4Addr;

use pnet::transport::TransportChannelType::Layer3;
use pnet::transport::TransportSender;
use pnet::packet::ip::IpNextHeaderProtocols;
use pnet::packet::ipv4::MutableIpv4Packet;

fn main() {
    let protocol = Layer3(IpNextHeaderProtocols::Icmp);
    let (mut tx, _) = pnet::transport::transport_channel(1024, protocol).unwrap();
    let addr_v4 = Ipv4Addr::new(127, 1, 2, 3);
    let mut vec: Vec<u8> = vec![0; 64];
    let mut vec2: Vec<u8> = vec![0; 64];
    let pkg = packet(&mut vec, addr_v4, false);
    let pkg2 = packet(&mut vec2, addr_v4, true);
    send(&mut tx, pkg, addr_v4);
    send(&mut tx, pkg2, addr_v4);
}

fn packet<'a>(vec: &'a mut Vec<u8>, addr: Ipv4Addr, flags: bool) -> MutableIpv4Packet<'a> {
    let mut packet = MutableIpv4Packet::new(vec).unwrap();
    packet.set_version(4);
    packet.set_header_length(5);
    packet.set_ttl(1);
    if flags {
        packet.set_flags(2);
    }
    packet.set_destination(addr);
    packet.set_total_length(64);
    packet.set_fragment_offset(0);
    packet
}

fn send(tx: &mut TransportSender, pkg: MutableIpv4Packet, addr: Ipv4Addr) {
    let l = pkg.get_total_length();
    match tx.send_to(pkg, pnet::util::IpAddr::V4(addr)) {
        Ok(n) => assert_eq!(n, l as usize),
        Err(e) => panic!("failed to send packet: {}", e),
    };
}

```

**Comment by mrmonday** [comment]
Thanks for the test case.

I'm not near an OS X machine right now, I'll try and take a look later today when I have access to one. When working at the transport layer, OSes have a nasty habit of modifying your packets before sending them, or switching endianness of certain fields (or, even, just pretending your packets never existed and not giving any kind of errors...) - it's possible there's a missed case somewhere.

I'd be happy to help mentor you if you'd like to look into the issue yourself?

**Comment by faern** [comment]
I have now inserted some debugging `println!` in `pnet::internal::send_to` just before the call to `libc::sendto`

I extract the flags and offset fields with this code:
```rust
let flags = buffer[6] >> 5; // Three highest bits of byte 7
let offset = ((buffer[6] as u16 & 0b11111) << 8) | buffer[7] as u16;
```
The output for the first packet (with `set_flags` not called):

    internal::sendto: flags: 0. offset: 0

Byte 7 & 8 in wireshark:

    00 00

The output for the second packet (with `set_flags` called):

    internal::sendto: flags: 2. offset: 0

Byte 7 & 8 in wireshark:

    00 40

So my conclusion is that pnet sends the correct bytes off to libc but something is changing the packet before it hits the wire. I'm not sure how to continue from here.

**Comment by mrmonday** [comment]
I'm going to summarise some of what's been said on IRC for posterity:

There are two main useful bits of documentation for raw sockets on OS X (that come to mind - I remember there being others, but I can't think of them specifically):
 * `man 4 ip`, under Raw IP Sockets - https://developer.apple.com/library/mac/documentation/Darwin/Reference/ManPages/man4/ip.4.html
 * This shiny web page: http://sock-raw.org/papers/sock_raw

There are at least three peculiarities with the way OS X handles IPv4 packets with raw sockets (possibly another with this bug):
 * The `fragment_offset` and `total_length` fields need to be specified in host endianness rather than network endianness like the other fields - this is handled with the `fixup_packet()` function in `src/transport.rs`.
 * Certain fields are modified by the OS before the packet hits the wire, in particular the `identification` and `checksum` fields - there is, to the best of my knowledge, nothing which can be done about this.
 * It doesn't play nicely with protocols already handled by the kernel - some details of this can be found on the sock_raw page linked above. This behaviour varies from "doesn't work at all" to "some subset of things work".

**Comment by mrmonday** [comment]
Fixed in pull #165
